### PR TITLE
[codex] Reduce pipeline redundancy with focused owners

### PR DESCRIPTION
**Pull request by zfifteen**
## Summary
- extract focused owners for scene validation, metadata resolution, voice-cache readiness, and render verification
- reduce duplicated shell logic in the main pipeline while preserving the canonical entrypoint
- add deterministic regression tests for the new seams and review-discovered edge cases

## Verification
- python3 scripts/test_scene_validator.py
- python3 scripts/test_build_scenes_validator_integration.py
- python3 scripts/test_scene_repair_invocation.py
- python3 scripts/test_resolve_scene_metadata.py
- python3 scripts/test_build_scenes_metadata_resolver_integration.py
- python3 scripts/test_state_record_error_integration.py
- python3 scripts/test_phase_progress_sentinel.py
- python3 scripts/test_voice_cache_integration.py
- python3 scripts/test_final_render_verification_integration.py
- python3 scripts/test_verify_scene_video.py
- python3 scripts/test_record_scene_rendered.py
- python3 scripts/test_contracts_vs_runtime.py
- bash -n scripts/build_video.sh

Supersedes closed PR #111 for a clean review surface.

### CODE REVIEW

File: scripts/build_video.sh [hunk 1]
```diff
@@ -2415,74 +2523,28 @@ PY
     exit 1
   fi
 
+  local voice_output_dir
+  voice_output_dir="$(voice_output_dir_abs)" || return 1
+
```
[the remark this answers is not in the document]
Comment (zfifteen) [reply scripts/build_video.sh:2528]: Verified this path in bc29d91: `voice_output_dir` is used later in `handle_final_render` to build `scene_audio` for the rendered-scene freshness check against cached narration audio. I left the runtime code intact and added an integration assertion so this use remains explicit.

File: scripts/record_scene_rendered.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    parser.add_argument("--scene-id", required=True)
+    parser.add_argument("--class-name", required=True)
+    args = parser.parse_args()
+
+    try:
+        record_scene_rendered(args.project_dir, args.scene_id, args.class_name)
+    except (OSError, ValueError, json.JSONDecodeError) as exc:
+        print(f"ERROR: {exc}")
+        return 1
+    return 0
```
[the remark this answers is not in the document]
Comment (zfifteen) [reply scripts/record_scene_rendered.py:99]: Fixed in bc29d91. `record_scene_rendered.py` now prints its error path to stderr, and `test_record_scene_rendered.py` asserts failure diagnostics are on stderr with empty stdout.

File: scripts/test_scene_validator.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    def test_comment_only_body_fails_before_pipeline_execution(self):
+        with tempfile.TemporaryDirectory() as temp_dir:
+            project_dir = Path(temp_dir)
+            scene_file = _write_scene(project_dir, "            # TODO")
+
+            result = _run_validator(scene_file)
+            self.assertEqual(result.returncode, 1)
+            payload = json.loads(result.stdout)
+            self.assertEqual(payload["first_failed_gate"], "python_syntax")
+
```
[the remark this answers is not in the document]
Comment (zfifteen) [reply scripts/test_scene_validator.py]: Addressed in 0a17d44. I verified the validator behavior before changing the test: the comment-only slot is inside the full scaffold `with self.voiceover(...)` block, and because comments do not count as executable statements, full-scene compilation fails first with `python_syntax` (`expected an indented block`). The patch renames the test to make that full-scaffold contract explicit and asserts the syntax-error summary so the expected gate is no longer ambiguous.

File: scripts/resolve_scene_metadata.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    if value is None:
+        return ""
+    if not isinstance(value, str) or not value.strip():
+        raise ValueError(f"scene[{scene_index}].{field} must be a non-empty string when present")
+    return value
+
+
+def resolve_scene_metadata(project_dir: Path, scene_index: int | None = None) -> dict[str, Any]:
+    state = _load_state(project_dir)
+    scenes = state.get("scenes") or []
```
[the remark this answers is not in the document]
Comment (zfifteen) [reply scripts/resolve_scene_metadata.py]: Fixed in a6c5c34. `resolve_scene_metadata()` now preserves the raw `scenes` value and only defaults to `[]` when the field is missing or `null`; falsey non-lists such as `{}` now fail fast with `project_state.json scenes must be a list`. Added regression coverage for missing/null scenes and the corrupted `{}` case.

File: scripts/build_video.sh [hunk 5]
```diff
@@ -1026,6 +1135,71 @@ validate_scene_semantics() {
...
+SCENE_VALIDATOR_REASON=""
+
+validate_scene_first_pass_with_owner() {
+  local scene_file="$1"
+  local validator_output parsed_output
+
+  SCENE_VALIDATOR_GATE=""
+  SCENE_VALIDATOR_REASON=""
+
+  validator_output="$(mktemp "${TMPDIR:-/tmp}/flaming-horse-scene-validator.XXXXXX.json")"
```
[the remark this answers is not in the document]
Comment (zfifteen) [reply scripts/build_video.sh]: Fixed in 464202a. The scene-validator temp file now uses a BSD/macOS-safe `mktemp` template that ends with the X placeholders: `flaming-horse-scene-validator-json.XXXXXX`. I also ran a targeted search to verify no remaining `mktemp ... XXXXXX.<suffix>` patterns in the runtime/test surfaces.

File: scripts/verify_scene_video.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+    ffprobe = shutil.which("ffprobe")
+    if not ffprobe:
+        return _result(
+            True,
+            video_path,
+            "ffprobe not found; skipped audio verification",
+            False,
+            None,
+        )
```
[the remark this answers is not in the document]
Comment (zfifteen) [reply scripts/verify_scene_video.py:59]: Fixed in 464202a. Added `test_ffprobe_missing_skips_audio_verification`, which constrains `PATH` to a temp bin containing only a `python3` symlink, forcing `shutil.which("ffprobe")` to return `None`. The test asserts `ok=true`, `audio_checked=false`, `audio_present=null`, and the stable skip reason.

File: scripts/record_scene_rendered.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    project_dir = project_dir.resolve()
+    state_file = project_dir / "project_state.json"
+    video_path = scene_video_path(project_dir, scene_id, class_name)
+    state = json.loads(state_file.read_text(encoding="utf-8"))
+    matching_scene = None
+    for scene in state.get("scenes", []):
+        if scene.get("id") == scene_id:
+            matching_scene = scene
+            break
+    if matching_scene is None:
```
[the remark this answers is not in the document]
Comment (zfifteen) [reply scripts/record_scene_rendered.py:64]: Fixed in 18c3d35. `record_scene_rendered()` now validates that `project_state.json` is an object, `scenes` is a list, and each scene entry is an object before iterating. Malformed state now fails through the existing CLI error path as a stable `ERROR:` message with no traceback. Added regression coverage for `scenes: null`, `scenes: {}`, and a non-object scene entry.

**Comment by zfifteen** [comment]
Closing this PR to open a fresh review surface for the same branch after multiple review rounds.
